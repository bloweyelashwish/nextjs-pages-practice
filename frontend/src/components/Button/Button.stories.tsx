import { Meta, StoryObj } from '@storybook/react';
import { expect } from '@storybook/jest';
import { screen, userEvent } from '@storybook/testing-library';

import { Button } from './Button';

const meta: Meta<typeof Button> = {
  title: 'Controls/Button',
  tags: ['autodocs'],
  component: Button,
  args: {
    children: 'Button text',
  },
};

export default meta;

type Story = StoryObj<typeof Button>;

export const PrimaryButton: Story = {
  play: async ({ args }) => {
    await userEvent.click(screen.getByRole('button'));
    await expect(args.onClick).toHaveBeenCalled();
  },
  args: {
    color: 'primary',
  },
};

export const SecondaryButton: Story = {
  ...PrimaryButton,
  args: {
    color: 'secondary',
  },
};

export const WarningButton: Story = {
  ...PrimaryButton,
  args: {
    color: 'warning',
  },
};

export const DangerButton: Story = {
  ...PrimaryButton,
  args: {
    color: 'danger',
  },
};
