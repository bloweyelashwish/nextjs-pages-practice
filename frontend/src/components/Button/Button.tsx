import { FC } from 'react';
import { css } from '@emotion/react';
import type { SerializedStyles } from '@emotion/react';
import styled from '@emotion/styled';
import type { MouseEvent } from 'react';

import type { AppTheme } from '~/styles/themes';
import { borderRadius, boxShadow, transition } from '~/styles/common';

export type ButtonColor = 'primary' | 'secondary' | 'danger' | 'warning';

export interface ButtonProps {
  children: string;
  color?: ButtonColor;
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
}

export const getColors = (
  theme: AppTheme,
  color?: ButtonColor
): SerializedStyles => {
  switch (color) {
    case 'secondary':
      return css`
        color: ${theme.font.regular};
      `;
    case 'primary':
    case 'danger':
      return css`
        background: ${theme.components[color]};
        color: ${theme.font.button};
      `;
    case 'warning':
      return css`
        background: ${theme.components[color]};
        color: ${theme.font.warning};
      `;
    default:
      return css``;
  }
};

export const Button = styled.button<ButtonProps>`
  all: unset;
  display: flex;
  justify-items: center;
  align-items: center;
  user-select: none;
  cursor: pointer;
  font-size: 1.6rem;
  padding: 1rem 3rem;
  outline: none;
  border: none;

  ${borderRadius()};
  ${transition()};
  ${({ color, theme }) => getColors(theme, color)};
  ${({ theme }) =>
    boxShadow(theme.components.shadow1, theme.components.shadow2)};

  &:hover {
    opacity: 0.75;
  }

  &:active {
    ${({ theme }) =>
      boxShadow(theme.components.shadow1, theme.components.shadow2, true)};
  }
`;

Button.defaultProps = {
  color: 'primary',
};

type DefinedButton = Omit<ButtonProps, 'color'>;

export const PrimaryButton: FC<DefinedButton> = (props) => (
  <Button color="primary" {...props} />
);

export const SecondaryButton: FC<DefinedButton> = (props) => (
  <Button color="secondary" {...props} />
);

export const DangerButton: FC<DefinedButton> = (props) => (
  <Button color="danger" {...props} />
);

export const WarningButton: FC<DefinedButton> = (props) => (
  <Button color="warning" {...props} />
);
