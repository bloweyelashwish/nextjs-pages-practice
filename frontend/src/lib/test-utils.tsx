import React from 'react';
import type { FC, PropsWithChildren, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from '@emotion/react';

import { Themes } from '~/styles/themes';

const ThemeWrapper: FC<PropsWithChildren> = ({ children }) => (
  <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
);

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: ThemeWrapper, ...options });

export * from '@testing-library/react';
export { customRender as render };
