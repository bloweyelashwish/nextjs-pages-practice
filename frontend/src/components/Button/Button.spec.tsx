import { render, screen } from '~/lib/test-utils';
import userEvent from '@testing-library/user-event';

import { Button } from './Button';

describe('Button test cases', () => {
  it('Button renders', () => {
    const clickHandler = jest.fn();
    const { asFragment } = render(
      <Button onClick={clickHandler}>Test button</Button>
    );

    expect(asFragment()).toMatchSnapshot();
  });
  it('onClick handler fires on user click', async () => {
    const clickHandler = jest.fn();
    render(<Button onClick={clickHandler}>Test button</Button>);
    const element = screen.getByRole('button');

    await userEvent.click(element);

    expect(clickHandler).toHaveBeenCalled();
  });
});
