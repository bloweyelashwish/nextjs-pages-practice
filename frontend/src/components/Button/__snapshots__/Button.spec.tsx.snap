// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button test cases Button renders 1`] = `
<DocumentFragment>
  <button
    class="css-1buu1e4"
    color="primary"
  >
    Test button
  </button>
</DocumentFragment>
`;
